### PR TITLE
MicroBit extension, Scratch Link first draft.

### PR DESCRIPTION
**Pull request by evhan55**
### Proposed Changes

This pull request contains a new MicroBit extension proposal, using the new Scratch Link protocol.  It includes **io** and **util** classes to support connecting to the new Scratch Link.  It does not yet support user-chosen devices, instead connecting to the first MicroBit discovered via Scratch Link.

Meant to be used for testing as we continue to develop a new user interface for device choosing.

### Reason for Changes

New Scratch 3.0 extension and device management.

### Test Coverage

All tests added are stubs for now, until the user interface details are further fleshed out.

### Notes and Questions for Reviewers:

1. `TODOs` left in there for GUI/PeripheralChooser progress
2. Had to change some of the ` != null` checks to make the linter pass.  @cwillisf , this _seems_ to not break anything, but wanted to point you to it.

### CODE REVIEW

File: src/util/jsonrpc.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    _sendRequest (method, params, id) {
+        const request = {
+            jsonrpc: '2.0',
+            method,
+            params
+        };
+
+        if (id !== null) {
```
Comment (evhan55) [inline src/util/jsonrpc.js]: Had to change this from 
```javascript
if (id != null)
```

File: src/util/jsonrpc.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            response.result = result || null;
+        }
+        this._sendMessage(response);
+    }
+
+    _handleResponse (json) {
+        const {result, error, id} = json;
+        const openRequest = this._openRequests[id];
+        delete this._openRequests[id];
+        if (error) {
```
Comment (evhan55) [inline src/util/jsonrpc.js]: Had to change this from:
```javascript
if (error != null)
```

File: src/util/jsonrpc.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            openRequest.reject(error);
+        } else {
+            openRequest.resolve(result);
+        }
+    }
+
+    _handleRequest (json) {
+        const {method, params, id} = json;
+        const rawResult = this.didReceiveCall(method, params);
+        if (id) {
```
Comment (evhan55) [inline src/util/jsonrpc.js]: Had to change this from:
```javascript
if (id != null)
```

File: src/util/jsonrpc-web-socket.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    _onSocketClose (e) {
+        log.info(`WS closed: ${e}`);
+    }
+
+    _onSocketError (e) {
+        log.info(`WS error: ${e}`);
+    }
+
+    _onSocketMessage (e) {
+        log.info(`Received message: ${e.data}`);
```
Comment (evhan55) [inline src/util/jsonrpc-web-socket.js]: This line gets called quite often if a device is connected and sending data, might need to remove or comment out, etc.

File: src/util/base64-util.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+class Base64Util {
+
+    /**
+     * Convert a base64 encoded string to a Uint8Array.
+     * @param {string} base64 - a base64 encoded string.
+     * @return {Uint8Array} - a decoded Uint8Array.
+     */
+    static base64ToUint8Array (base64) {
+        const binaryString = window.atob(base64);
```
Comment (evhan55) [inline src/util/base64-util.js]: Not sure about this call to `window`.

File: src/util/base64-util.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        return array;
+    }
+
+    /**
+     * Convert a Uint8Array to a base64 encoded string.
+     * @param {Uint8Array} array - the array to convert.
+     * @return {string} - the base64 encoded string.
+     */
+    static uint8ArrayToBase64 (array) {
+        const base64 = window.btoa(String.fromCharCode.apply(null, array));
```
Comment (evhan55) [inline src/util/base64-util.js]: Not sure about this call to `window`.

File: src/io/scratchBT.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
+const JSONRPCWebSocket = require('../util/jsonrpc');
+const log = require('../util/log');
+
+const ScratchLinkWebSocket = 'ws://localhost:20110/scratch/ble';
```
Comment (evhan55) [inline src/io/scratchBT.js]: Ah whoops, this must probably be `ws://localhost:20110/scratch/bt`

File: src/io/scratchBT.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
+const JSONRPCWebSocket = require('../util/jsonrpc');
+
+const ScratchLinkWebSocket = 'ws://localhost:20110/scratch/bt';
+
+class ScratchBT extends JSONRPCWebSocket {
```
Comment (evhan55) [inline src/io/scratchBT.js:5]: Presumably, this file will fill out a bit once we work on a Bluetooth Classic device extension, like EV3 which is coming up

**Comment by evhan55** [comment]
I see some functions in the Microbit extension I might want to rename, now that I am working on the EV3 one as well, but I will leave this PR as is and keep a mental note to change them in a later PR

File: src/extensions/scratch3_microbit/index.js [hunk 9]
```diff
@@ -144,31 +182,38 @@ class MicroBit {
...
+    _onBLEConnect () {
+        const callback = this._processBLEData.bind(this);
+        this._ble.read(BLEUUID.service, BLEUUID.rxChar, true, callback);
+    }
+
+    /**
+     * @param {string} e - Error from BLE session.
+     */
+    _onBLEError (e) {
+        log.error(`BLE error: ${e}`);
```
Comment (evhan55) [inline src/extensions/scratch3_microbit/index.js:207]: I think this error needs to be stringified probably

File: src/io/scratchBLE.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        });
+    }
+
+    /**
+     * Request a device with the device options and optional gui options.
+     * @param {object} deviceOptions - list of device guiOptions.
+     * @param {object} onConnect - on connect callback.
+     * @param {object} onError - on error callbackk.
+     */
+    requestDevice (deviceOptions, onConnect, onError) {
```
Comment (evhan55) [inline src/io/scratchBLE.js:34]: This (or someone else, the peripheral chooser?) should probably handle what happens when 'discover' doesn't find any devices after a while

**Review by ericrosenbaum — APPROVED** [review]
yay!

**Pull request merged**